File: submissions/python3/16.py
class Solution:
    def threeSumClosest(self, nums, target):
        """
        :type nums: List[int]
        :type target: int
        :rtype: int
        """
        nums.sort()
        ret, minAbs = 0, sys.maxsize
        for i in range(len(nums)-2):
            j, k = i + 1, len(nums) - 1
            minus = target - nums[i]
            while j < k:
                sum = nums[j] + nums[k]
                abso = abs(minus - sum)
                if minAbs > abso:
                    minAbs = abso
                    ret = sum + nums[i]
                if minAbs == 0:
                    return ret
                if sum > minus:
                    k -= 1
                else:
                    j += 1
        return ret

File: LeetCode/cpp/16.cpp
class Solution {
public:
    int threeSumClosest(vector<int> &nums, int target) {
        auto size = nums.size();
        sort(nums.begin(), nums.end());
        int minAbs = INT_MAX, ret = 0;
        for (int i = 0; i < size - 2; ++i) {
            int x = i + 1, y = size - 1;
            while (x < y) {
                int sum = nums[x] + nums[y], tar = target - nums[i];
                int num = abs(sum - tar);
                if (minAbs > num) {
                    minAbs = num;
                    ret = nums[x] + nums[y] + nums[i];
                }
                if (minAbs == 0)
                    return ret;
                if (sum < tar)
                    ++x;
                else
                    --y;
            }
        }
        return ret;
    }
};
